components/resource_control/src/future.rs

```diff
@@ -44,3 +124,312 @@ impl<F: Future> Future for ControlledFuture<F> {
         res
     }
 }
+
+// `TrackIoStateFuture` is used to update the thread local variable
+// `THREAD_TOTAL_IO_BYTES`. it should be used pair with `LimitedFuture` so the
+// `THREAD_TOTAL_IO_BYTES` state is valid.
+#[pin_project]
+pub struct TrackIoStateFuture<F> {
+    #[pin]
+    f: F,
+}
+
+impl<F: Future> TrackIoStateFuture<F> {
+    #[allow(dead_code)]
+    pub fn new(f: F) -> Self {
+        Self { f }
+    }
+}
+
+impl<F: Future> Future for TrackIoStateFuture<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        THREAD_TOTAL_IO_BYTES.with(|s| {
```

Comment:
It confuses me that both `TrackIoStateFuture` and `LimitedFuture` will change the LocalIoState state, I think the logic of state change is hard to understand.

Reply:
If we don't wrap all other future tasks with `TrackIoStateFuture`, we need to always call `get_threshold_io_stats ` before and after poll  `LimitedFuture` to calculate the actual io bytes. `get_threshold_io_stats` cost about 10us which is not very fast. So this is used to optimize the performance.